server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java

```diff
@@ -544,4 +530,80 @@ private void processNormalizerFactory(
         }
         normalizers.put(name, normalizer);
     }
+
+    /**
+     * Create an new IndexAnalyzer instance based on the existing one. Analyzers that are in {@link AnalysisMode#SEARCH_TIME} are tried to
+     * be reloaded. All other analyzers are reused from the old {@link IndexAnalyzers} instance.
+     */
+    public IndexAnalyzers rebuildIndexAnalyzers(IndexAnalyzers indexAnalyzers, IndexSettings indexSettings) throws IOException {
```

Comment:
This and the following method are the main places where reloading happens

Reply:
I wonder if we should collect the list of filters, char_filters and tokenizers that need a rebuild in all search analyzer first. The current solution seems to rebuild the filters once per analyzer instead of only once.

Reply:
Ok I think I understand why I got confused. Synonym token filters are special because they delay the build of the synonym map until all other filters are known. This is the reason why your code is working even though you're not rebuilding the token and char filter factories, the synonym map is rebuild when the custom analyzer is finally built. 
However if we want to make the update work on other type of filters we'll need to rebuild the factories first and then use these new factories to build the custom analyzers.